components/omega/src/ocn/Eos.h

```diff
@@ -261,11 +262,266 @@ class LinearEos {
       const I4 KStart = KChunk * VecLength;
       for (int KVec = 0; KVec < VecLength; ++KVec) {
          const I4 K = KStart + KVec;
+         if (K >= NVertLayers)
+            continue;
          SpecVol(ICell, K) =
              1.0_Real / (RhoT0S0 + (DRhodT * ConservTemp(ICell, K) +
                                     DRhodS * AbsSalinity(ICell, K)));
       }
    }
+
+ private:
+   I4 NVertLayers;
+};
+
+/// Functor for calculating the Brunt-Vaisala frequency using TEOS-10
+class Teos10BruntVaisalaFreq {
+ public:
+   /// Constructor for BruntVaisalaFreq
+   Teos10BruntVaisalaFreq(const VertCoord *VCoord);
+
+   //   The functor takes the full arrays of Brunt-Vaisala frequency (inout),
+   //   the indix ICell, and the ocean tracers (conservative) temperature,
+   //   (absolute) salinity, pressure, specific volume as inputs, and outputs
+   //   the Brunt-Vaisala frequency.
+   KOKKOS_FUNCTION void operator()(Array2DReal BruntVaisalaFreq, I4 ICell,
+                                   I4 KChunk, const Array2DReal &ConservTemp,
+                                   const Array2DReal &AbsSalinity,
+                                   const Array2DReal &Pressure,
+                                   const Array2DReal &SpecVol) const {
+      Real Gravity    = 9.80616;
+      Real Db2Pa      = 1.0e4;
+      const I4 KStart = KChunk * VecLength;
+      for (int KVec = 0; KVec < VecLength; ++KVec) {
+         const I4 K = KStart + KVec;
+         if (K >= NVertLayers)
+            continue;
+         else if (K == 0) {
+            // No Brunt-Vaisala frequency at surface
+            BruntVaisalaFreq(ICell, K) = 0.0_Real;
+         } else {
+            // Calculate Brunt-Vaisala frequency
+            Real CtInt =
+                0.5_Real * (ConservTemp(ICell, K) + ConservTemp(ICell, K - 1));
+            Real SaInt =
+                0.5_Real * (AbsSalinity(ICell, K) + AbsSalinity(ICell, K - 1));
+            Real PInt =
+                0.5_Real * (Pressure(ICell, K) + Pressure(ICell, K - 1));
+            Real SpInt = 0.5_Real * (SpecVol(ICell, K) + SpecVol(ICell, K - 1));
+            Real AlphaInt = calcAlpha(SaInt, CtInt, PInt, SpInt);
+            Real BetaInt  = calcBeta(SaInt, CtInt, PInt, SpInt);
+            Real DSa      = AbsSalinity(ICell, K) - AbsSalinity(ICell, K - 1);
+            Real DCt      = ConservTemp(ICell, K) - ConservTemp(ICell, K - 1);
+            Real DP       = Pressure(ICell, K) - Pressure(ICell, K - 1);
+
+            BruntVaisalaFreq(ICell, K) = Gravity * Gravity *
+                                         ((1.0_Real / SpInt) / (Db2Pa * DP)) *
```

Comment:
It's a minor thing but wouldn't it make sense to keep the straightforward expression that match the [teos-10 doc](https://www.teos-10.org/pubs/gsw/html/gsw_Nsquared.html)?
```suggestion
                                         (BetaInt * DSa - AlphaInt * DCt)/ (SpInt * Db2Pa * DP)
```